dist-tracing/src/telemetry_layer.rs

```diff
@@ -0,0 +1,445 @@
+        }
+    }
+
+    pub fn new() -> Self {
+        let instance_id = rand::thread_rng().gen();
+        let registry = RwLock::new(HashMap::new());
+
+        TraceCtxRegistry {
+            instance_id,
+            registry,
+        }
+    }
+}
+
+impl<T> TelemetryLayer<T> {
+    pub fn new(service_name: String, telemetry: T) -> Self {
+        let span_data = TraceCtxRegistry::new();
+
+        TelemetryLayer {
+            service_name,
+            telemetry,
+            span_data,
+        }
+    }
+}
+
+impl<
+        S,
+        V: tracing::field::Visit + Default + Send + Sync + 'static,
+        T: 'static + Telemetry<Visitor = V>,
+    > Layer<S> for TelemetryLayer<T>
+where
+    S: Subscriber + for<'a> registry::LookupSpan<'a>,
+{
+    fn new_span(&self, attrs: &Attributes, id: &Id, ctx: Context<S>) {
+        let span = ctx.span(id).expect("span data not found during new_span");
+        let mut extensions_mut = span.extensions_mut();
+        extensions_mut.insert(SpanInitAt::new());
+
+        let mut visitor: V = Default::default();
+        attrs.record(&mut visitor);
+        extensions_mut.insert::<V>(visitor);
+    }
+
+    fn on_record(&self, id: &Id, values: &Record, ctx: Context<S>) {
+        let span = ctx.span(id).expect("span data not found during on_record");
+        let mut extensions_mut = span.extensions_mut();
+        let visitor: &mut V = extensions_mut
+            .get_mut()
+            .expect("fields extension not found during on_record");
+        values.record(visitor);
+    }
+
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        let parent_id = if let Some(parent_id) = event.parent() {
+            // explicit parent
+            Some(parent_id.clone())
+        } else if event.is_root() {
+            // don't bother checking thread local if span is explicitly root according to this fn
+            None
+        } else if let Some(parent_id) = ctx.current_span().id() {
+            // implicit parent from threadlocal ctx
+            Some(parent_id.clone())
+        } else {
+            // no parent span, thus this is a root span
+            None
+        };
+
+        match parent_id {
+            None => {} // not part of a trace, don't bother recording via honeycomb
+            Some(parent_id) => {
+                let initialized_at = Utc::now();
+
+                let mut visitor = Default::default();
+                event.record(&mut visitor);
+
+                // TODO: dedup
+                let iter = itertools::unfold(Some(parent_id.clone()), |st| match st {
+                    Some(target_id) => {
+                        let res = ctx
+                            .span(target_id)
+                            .expect("span data not found during eval_ctx");
+                        *st = res.parent().map(|x| x.id());
+                        Some(res)
+                    }
+                    None => None,
+                });
```

Comment:
seems like you could just use the existing [`SpanRef::parents`](https://docs.rs/tracing-subscriber/0.2.0/tracing_subscriber/registry/struct.SpanRef.html#method.parents) iterator for this:
```rust
let iter = ctx.span(id).iter().flat_map(|span| span.parents());
```
or something